File: typescript/src/n0383_ransom_note.ts
export function canConstruct(ransom_note: string, magazine: string): boolean {
  const counts: number[] = Array(26).fill(0);
  for (const c of magazine) {
    const index = c.charCodeAt(0) - 97;
    counts[index]! += 1;
  }
  for (const c of ransom_note) {
    const index = c.charCodeAt(0) - 97;
    if (counts[index] === 0) {
      return false;
    }
    counts[index]! -= 1;
  }
  return true;
}

File: rust/src/n0383_ransom_note.rs
pub fn can_construct(ransom_note: &str, magazine: &str) -> bool {
    let mut counts = [0; 26];
    for b in magazine.bytes() {
        let index = (b - b'a') as usize;
        counts[index] += 1;
    }
    for b in ransom_note.bytes() {
        let index = (b - b'a') as usize;
        if counts[index] == 0 {
            return false;
        }
        counts[index] -= 1;
    }
    true
}
